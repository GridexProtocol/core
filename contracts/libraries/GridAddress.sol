// SPDX-License-Identifier: GPL-2.0-or-later
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/Create2.sol";

library GridAddress {
    bytes32 internal constant GRID_BYTES_CODE_HASH = 0xa5d3f2e5878f7c232384a9d32cf774cad0591c5024b26e6fccb47ecc09433920;

    struct GridKey {
        address token0;
        address token1;
        int24 resolution;
    }

    function gridKey(address tokenA, address tokenB, int24 resolution) internal pure returns (GridKey memory) {
        if (tokenA > tokenB) (tokenA, tokenB) = (tokenB, tokenA);

        return GridKey(tokenA, tokenB, resolution);
    }

    function computeAddress(address gridFactory, GridKey memory key) internal pure returns (address) {
        require(key.token0 < key.token1);
        return
            Create2.computeAddress(
                keccak256(abi.encode(key.token0, key.token1, key.resolution)),
                GRID_BYTES_CODE_HASH,
                gridFactory
            );
    }
}
