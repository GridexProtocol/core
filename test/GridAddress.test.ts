import {ethers} from "hardhat";
import {expect} from "./shared/expect";
import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {GridAddressTest} from "../typechain-types";
import {Resolution} from "./shared/util";
import {computeAddress, GRID_BYTES_CODE_HASH} from "./shared/GridAddress";

describe("GridAddress", () => {
    const tokenAddresses = ["0x1000000000000000000000000000000000000000", "0x2000000000000000000000000000000000000000"];
    const gridFactoryAddress = "0x9000000000000000000000000000000000000000";

    async function deployFixture() {
        const [signer, otherAccount] = await ethers.getSigners();
        const gridAddressFactory = await ethers.getContractFactory("GridAddressTest", signer);
        const gridAddress = (await gridAddressFactory.deploy()) as GridAddressTest;
        return {gridAddress, signer, otherAccount};
    }

    describe("#GRID_BYTES_CODE_HASH", () => {
        it(`grid bytes code hash should equal to ${GRID_BYTES_CODE_HASH}`, async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            expect(await gridAddress.GRID_BYTES_CODE_HASH()).to.equal(GRID_BYTES_CODE_HASH);
        });
    });

    describe("#gridKey", () => {
        it("should sort by order", async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            const {key} = await gridAddress.gridKey(tokenAddresses[0], tokenAddresses[1], Resolution.LOW);
            expect(key.token0).to.equal(tokenAddresses[0]);
            expect(key.token1).to.equal(tokenAddresses[1]);
            expect(key.resolution).to.equal(Resolution.LOW);
        });

        it("should sort by order", async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            const {key} = await gridAddress.gridKey(tokenAddresses[1], tokenAddresses[0], Resolution.HIGH);
            expect(key.token0).to.equal(tokenAddresses[0]);
            expect(key.token1).to.equal(tokenAddresses[1]);
            expect(key.resolution).to.equal(Resolution.HIGH);
        });

        it("gas used snapshot", async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            const {gasUsed} = await gridAddress.gridKey(tokenAddresses[1], tokenAddresses[0], Resolution.LOW);
            expect(gasUsed.toNumber()).toMatchSnapshot();
        });
    });

    describe("#computeAddress", () => {
        it("should revert if token0 > token1", async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            await expect(
                gridAddress.computeAddress(gridFactoryAddress, {
                    token0: tokenAddresses[1],
                    token1: tokenAddresses[0],
                    resolution: Resolution.MEDIUM,
                })
            ).to.reverted;
        });

        it("should equal on chain and off chain", async () => {
            const {gridAddress} = await loadFixture(deployFixture);
            const onChainAddress = await gridAddress.computeAddress(gridFactoryAddress, {
                token0: tokenAddresses[0],
                token1: tokenAddresses[1],
                resolution: Resolution.MEDIUM,
            });
            const offChainAddress = await computeAddress(
                gridFactoryAddress,
                tokenAddresses[0],
                tokenAddresses[1],
                Resolution.MEDIUM
            );
            expect(onChainAddress).to.equal(offChainAddress);
        });
    });
});
