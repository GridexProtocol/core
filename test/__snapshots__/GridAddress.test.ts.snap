// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridAddress #gridKey gas used snapshot 1`] = `285`;
